README
NAME
    DBIx::Class::InflateColumn::ISBN - Auto-create Business::ISBN objects
    from columns.

VERSION
    Version 0.03000

SYNOPSIS
    Load this component and declare columns as ISBNs with the appropriate
    format.

        package Library;
        __PACKAGE__->load_components(qw/InflateColumn::ISBN Core/);
        __PACKAGE__->add_columns(
            isbn => {
                data_type => 'varchar',
                size => 13,
                is_nullable => 0,
                is_isbn => 1,
            }
        );

    It has to be a varchar rather than a simple integer given that it is
    possible for ISBNs to contain the character X. Old style ISBNs are 10
    characters, not including hyphens, but new style ones are 13 characters.

    Then you can treat the specified column as a Business::ISBN object.

        print 'ISBN: ', $book->isbn->as_string;
        print 'Publisher code: ', $book->isbn->publisher_code;

METHODS
  isbn_class
    Arguments: $class

    Gets/sets the address class that the columns should be inflated into.
    The default class is Business::ISBN and only that is currently
    supported.

  register_column
    Chains with "register_column" in DBIx::Class::Row, and sets up ISBN
    columns appropriately. This would not normally be called directly by end
    users.

AUTHOR
    K. J. Cheetham "<jamie @ shadowcatsystems.co.uk>"

BUGS
    Please report any bugs or feature requests to
    "bug-dbix-class-inflatecolumn-ISBN at rt.cpan.org", or through the web
    interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=DBIx-Class-InflateColumn
    -ISBN>. I will be notified, and then you'll automatically be notified of
    progress on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc DBIx::Class::InflateColumn::ISBN

    You can also look for information at:

    * AnnoCPAN: Annotated CPAN documentation
        <http://annocpan.org/dist/DBIx-Class-InflateColumn-ISBN>

    * CPAN Ratings
        <http://cpanratings.perl.org/d/DBIx-Class-InflateColumn-ISBN>

    * RT: CPAN's request tracker
        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=DBIx-Class-InflateColumn-I
        SBN>

    * Search CPAN
        <http://search.cpan.org/dist/DBIx-Class-InflateColumn-ISBN>

SEE ALSO
    Business::ISBN

    DBIx::Class::InflateColumn

    WWW::Scraper::ISBN

COPYRIGHT & LICENSE
    Copyright 2007 K. J. Cheetham, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for DBIx::Class::InflateColumn::ISBN

0.03000 Tue May 29 17:50:28 2007
    - Tweaked Makefile.PL and TODO files.
    - Nuked warnings from the tests.
    - Added workaround for 'validation error checked' test.

0.02000 Fri May 25 15:41:44 2007
    - Started using longer version numbers.
    - Fixed bug in deflate method for Business::ISBN 1.x.
    - Minor POD tweaks.

0.01 Fri May 25 15:14:08 2007
    - First version, released on an unsuspecting world.

0.00 Thu May 24 21:55:43 2007
    - First version, not released, just a placeholder.


t/lib/DBICTest.pm
package # hide from PAUSE 
    DBICTest;

use strict;
use warnings;
use DBICTest::Schema;

=head1 NAME

DBICTest - Library to be used by DBIx::Class test scripts.

=head1 SYNOPSIS

  use lib qw(t/lib);
  use DBICTest;
  use Test::More;
  
  my $schema = DBICTest->init_schema();

=head1 DESCRIPTION

This module provides the basic utilities to write tests against 
DBIx::Class.

=head1 METHODS

=head2 init_schema

  my $schema = DBICTest->init_schema(
    no_deploy=>1,
    no_populate=>1,
  );

This method removes the test SQLite database in t/var/DBIxClass.db 
and then creates a new, empty database.

This method will call deploy_schema() by default, unless the 
no_deploy flag is set.

Also, by default, this method will call populate_schema() by 
default, unless the no_deploy or no_populate flags are set.

=cut

sub init_schema {
    my $self = shift;
    my %args = @_;
    my $db_file = "t/var/test.db";

    unlink($db_file) if -e $db_file;
    unlink($db_file . "-journal") if -e $db_file . "-journal";
    mkdir("t/var") unless -d "t/var";

    my $dsn = $ENV{"DBICTEST_DSN"} || "dbi:SQLite:${db_file}";
    my $dbuser = $ENV{"DBICTEST_DBUSER"} || '';
    my $dbpass = $ENV{"DBICTEST_DBPASS"} || '';

   my $schema = DBICTest::Schema->compose_connection('DBICTest' => $dsn, $dbuser, $dbpass);
    $schema->storage->on_connect_do(['PRAGMA synchronous = OFF']);
    if ( !$args{no_deploy} ) {
        __PACKAGE__->deploy_schema( $schema );
        __PACKAGE__->populate_schema( $schema ) if( !$args{no_populate} );
    }
    return $schema;
}

=head2 deploy_schema

  DBICTest->deploy_schema( $schema );

This method does one of two things to the schema.  It can either call 
the experimental $schema->deploy() if the DBICTEST_SQLT_DEPLOY environment 
variable is set, otherwise the default is to read in the t/lib/sqlite.sql 
file and execute the SQL within. Either way you end up with a fresh set 
of tables for testing.

=cut

sub deploy_schema {
    my $self = shift;
    my $schema = shift;

    if ($ENV{"DBICTEST_SQLT_DEPLOY"}) {
        return $schema->deploy();
    } else {
        open IN, "t/lib/sqlite.sql";
        my $sql;
        { local $/ = undef; $sql = <IN>; }
        close IN;
        ($schema->storage->dbh->do($_) || print "Error on SQL: $_\n") for split(/;\n/, $sql);
    }
}

=head2 populate_schema

  DBICTest->populate_schema( $schema );

After you deploy your schema you can use this method to populate 
the tables with test data.

=cut

sub populate_schema {
    my $self = shift;
    my $schema = shift;

    $schema->populate('Library', [
        [ qw/id book isbn/ ],
        [ 1, 'HTML for the World Wide Web: With XHTML and CSS', '0321430840' ],
        [ 2, 'Final Fantasy X Official Strategy Guide', '190351133X' ],
    ]);

}

1;


t/lib/DBICTest/Schema/Library.pm
package # hide from PAUSE
    DBICTest::Schema::Library;

use base qw/DBIx::Class/;

__PACKAGE__->load_components(qw/InflateColumn::ISBN Core/);
__PACKAGE__->table('library');

__PACKAGE__->add_columns(
    id => {
        data_type   => 'integer',
        is_nullable => 0,
    },
    book => {
        data_type   => 'text',
        is_nullable => 0,
    },
    isbn => {
        data_type   => 'varchar',
        size        => 13,
        is_nullable => 0,
        is_isbn     => 1,
    }
);

__PACKAGE__->set_primary_key('id');
__PACKAGE__->add_unique_constraint(isbn => [ qw/isbn/ ]);

1;


lib/DBIx/Class/InflateColumn/ISBN.pm
package DBIx::Class::InflateColumn::ISBN;

use warnings;
use strict;

our $VERSION = '0.03000';

use base qw/DBIx::Class/;
__PACKAGE__->mk_classdata('isbn_class');
__PACKAGE__->isbn_class('Business::ISBN');

=head1 NAME

DBIx::Class::InflateColumn::ISBN - Auto-create Business::ISBN objects from columns.

=head1 VERSION

Version 0.03000

=head1 SYNOPSIS

Load this component and declare columns as ISBNs with the appropriate format.

    package Library;
    __PACKAGE__->load_components(qw/InflateColumn::ISBN Core/);
    __PACKAGE__->add_columns(
        isbn => {
            data_type => 'varchar',
            size => 13,
            is_nullable => 0,
            is_isbn => 1,
        }
    );

It has to be a varchar rather than a simple integer given that it is possible for
ISBNs to contain the character X. Old style ISBNs are 10 characters, not including
hyphens, but new style ones are 13 characters.

Then you can treat the specified column as a Business::ISBN object.

    print 'ISBN: ', $book->isbn->as_string;
    print 'Publisher code: ', $book->isbn->publisher_code;

=head1 METHODS

=head2 isbn_class

=over

=item Arguments: $class

=back

Gets/sets the address class that the columns should be inflated into.
The default class is Business::ISBN and only that is currently supported.

=head2 register_column

Chains with L<DBIx::Class::Row/register_column>, and sets up ISBN columns
appropriately. This would not normally be called directly by end users.

=cut

sub register_column {
    my ($self, $column, $info, @rest) = @_;
    $self->next::method($column, $info, @rest);

    return unless defined $info->{'is_isbn'};

    warn "Datatype is too small" if ($info->{'size'} && $info->{'size'} < 10);
    warn "Datatype must not be integer" if ($info->{'data_type'} eq 'integer');

    my $isbn_class = $info->{'isbn_class'} || $self->isbn_class || 'Business::ISBN';

    eval "use $isbn_class";
    $self->throw_exception("Error loading $isbn_class: $@") if $@;

    $self->inflate_column(
        $column => {
            inflate => sub { return $isbn_class->new(sprintf("%010s", shift)); },
            deflate => sub { return shift->isbn; },
        }
    );
}

=head1 AUTHOR

K. J. Cheetham C<< <jamie @ shadowcatsystems.co.uk> >>

=head1 BUGS

Please report any bugs or feature requests to
C<bug-dbix-class-inflatecolumn-ISBN at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=DBIx-Class-InflateColumn-ISBN>.
I will be notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc DBIx::Class::InflateColumn::ISBN

You can also look for information at:

=over 4

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/DBIx-Class-InflateColumn-ISBN>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/DBIx-Class-InflateColumn-ISBN>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=DBIx-Class-InflateColumn-ISBN>

=item * Search CPAN

L<http://search.cpan.org/dist/DBIx-Class-InflateColumn-ISBN>

=back

=head1 SEE ALSO

L<Business::ISBN>

L<DBIx::Class::InflateColumn>

L<WWW::Scraper::ISBN>

=head1 COPYRIGHT & LICENSE

Copyright 2007 K. J. Cheetham, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of DBIx::Class::InflateColumn::ISBN
